[Media Library] Update parameter key for number of items to fetch to be `per_page`

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14432
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

When logging in with application password, the Media Library only fetches 10 images regardless. From investigating and comparing with Android, the API request in iOS uses `number` parameter.

According to the API doc, `number` parameter does not exist and instead `per_page` is needed: https://developer.wordpress.org/rest-api/reference/media/#arguments

Android also uses `per_page` and thus do not see this issue:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/cc17141a0e910d5b808ba6bbfb4b7d393d3517a0/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpapi/media/BaseWPV2MediaRestClient.kt#L231

This PR updates the parameter to the correct one.

The above should also apply the same for Jetpack CP site.

For full Jetpack site, the parameter `number` should still be used, and this PR differentiates between the two usage.

Reference:
1. .org API https://developer.wordpress.org/rest-api/reference/media/#arguments
2. .com API https://developer.wordpress.com/docs/api/1.2/get/sites/%24site/media/


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

**a. No Jetpack case (Application Password)**
1. Have a .org site with no Jetpack, 
3. Upload more than 25 images in wp-admin Media > Library,
4. In the app, login to site with application password,
5. Go to Products > Add Product > Add Product Image > WordPress Media Library
6. Ensure all the images are loaded, not just 10.

**b. Jetpack CP site case**
1. Have a .org site with no Jetpack but with a plugin with Jetpack CP (I use "Blaze Ads"), 
2. Setup the Jetpack CP plugin and connect it to your WordPress.com account,
3. Upload more than 25 images in wp-admin Media > Library,
4. In the app, login to site with WordPress.com account,
5. Go to Products > Add Product > Add Product Image > WordPress Media Library
7. Ensure all the images are loaded, not just 10.


**c. WordPress.com / full Jetpack site case**
1. Have a .org site with Jetpack plugin, 
2. Setup the Jetpack plugin and connect it to your WordPress.com account,
3. Upload more than 25 images in wp-admin Media > Library,
4. In the app, login to site with WordPress.com account,
5. Go to Products > Add Product > Add Product Image > WordPress Media Library
6. Ensure all the images are loaded, not just 10.


**New: "media_type" fix test**
1. Have a .org site with no Jetpack, 
3. Upload some images and at least one video to media library,
5. In the app, login to site with application password,
6. Go to Products > Add Product > Add Product Image > WordPress Media Library
7. Ensure all the images are loaded, but the video is not shown

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested this on three type of sites above, on iPhone simulator with iOS 18.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
Update parameter key for number of items to fetch to be `per_page`

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Update release notes.

selanthiraiyan commented on Networking/Networking/Remote/MediaRemote.swift:
```diff
@@ -294,7 +294,7 @@ public extension MediaRemote {
     private enum ParameterKey {
         static let pageNumber: String = "page"
-        static let pageSize: String   = "number"
+        static let pageSize: String   = "per_page"
```
Thanks for working on this, @hafizrahman! I will wait for @hichamboushaba to review this.

---

I am jumping in as this parameter key change will affect both `loadMediaLibrary` and `loadMediaLibraryFromWordPressSite` methods in this `MediaRemote` file. 

The `number` parameter key is from the following documentation. https://developer.wordpress.com/docs/api/1.2/get/sites/%24site/media/

Please make sure to test the following login cases and update the PR testing instructions accordingly.

1. Jetpack
2. JCP
3. No Jetpack. (Application password)

I am afraid that one of the above cases could be broken/affected by this change. 

---
Nice to have: It would be nice to add a unit test to ensure that the parameter key is correct.

hafizrahman replied:
Oh good catch, I tested in WordPress.com console and indeed it wants `number`. Let me refactor this.

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/14444/commits/7215568fbe05f767eebc8710a371eceb86007256

Good for another review.

hafizrahman committed:
Differentiate parameter between dotcom API and dotorg API.

hichamboushaba commented:
@hafizrahman just sharing this [bug](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3109#discussion_r1844958555) that I identified in Android last week, it's similar to this one, we use the `mime_type` parameter for all implementations, while it should be `media_type` for `loadMediaLibraryFromWordPressSite`, do you think it makes sense to fix it on the same PR?

hafizrahman commented:
@hichamboushaba thanks for the ping! I'll have it added here.

hafizrahman committed:
Merge branch 'trunk' into issue/14432-media-library-missing-images

hafizrahman committed:
Update .org API to use media_type instead of mime_type

Issue originally identified in Android https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3109#discussion_r1844958555

hafizrahman committed:
Update comments

hafizrahman commented:
Added `media_type` fix on https://github.com/woocommerce/woocommerce-ios/pull/14444/commits/02bd74d9967b94e677b427694b36851bb6f8011a and updated test instruction at the top with this scenario.

selanthiraiyan reviewed (commented):
Thanks for the updates, Hafiz!

> b. Jetpack CP site case

Unfortunately, the above case doesn't work. I used Blaze Ads plugin to connect my WPCOM account with the JN site. 

While trying to load images from the media library, I end up seeing the following error alert. 

<img src="https://github.com/user-attachments/assets/71cd19a9-bb4b-41ca-a001-41e83803c058" width="320"/>

To debug this,

I looked into proxyman, and I see that I do receive a success response from the store for the following request. 

`https://public-api.wordpress.com/wp/v2/sites/1234/media?_fields=id%2Cdate_gmt%2Cslug%2Cmime_type%2Csource_url%2Calt_text%2Cmedia_details%2Ctitle&media_type=image&page=1&per_page=25`

I looked into Xcode logs, and I see the following decoding error.

```
<> Mapping Error: typeMismatch(Swift.Dictionary<Swift.String, Any>, Swift.DecodingError.Context(codingPath: [_CodingKey(stringValue: "Index 6", intValue: 6), CodingKeys(stringValue: "media_details", intValue: nil), CodingKeys(stringValue: "sizes", intValue: nil)], debugDescription: "Expected to decode Dictionary<String, Any> but found an array instead.", underlyingError: nil))
```

hichamboushaba commented:
> I looked into Xcode logs, and I see the following decoding error.
>
> ```
> <> Mapping Error: typeMismatch(Swift.Dictionary<Swift.String, Any>, Swift.DecodingError.Context(codingPath: [_CodingKey(stringValue: "Index 6", intValue: 6), CodingKeys(stringValue: "media_details", intValue: nil), CodingKeys(stringValue: "sizes", intValue: nil)], debugDescription: "Expected to decode Dictionary<String, Any> but found an array instead.", underlyingError: nil))
> ```

This looks like the `media_details.sizes` property is returning a dictionary instead of the expected object, I had this happen in my tests when I worked on this in Android, but it was only for non-image types, which was happening because we weren't passing the correct argument `media_type` and we were using the `mime_type` one.
(This issue shouldn't be because of the PR's changes, and should happen also on trunk I think)

I tried to replicate this, but I couldn't on a test site that I have right now. It's the EOD for me, but I'll try to take a look at this if needed tomorrow, if it's an issue then it would impact Android as well.

selanthiraiyan commented:
Thanks for taking a look, @hichamboushaba! 

> I tried to replicate this, but I couldn't on a test site that I have right now.

I have sent invites to my JN store to both you and @hafizrahman. (To your A8C emails)

hichamboushaba commented:
Using your site @selanthiraiyan, I can see that an image item has the sizes property empty, and when using the WPCom endpoint, this item is returned as an array instead of an empty object (I think this is something that happens a lot with WordPress, in FluxC we have an old [class](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/JsonObjectOrEmptyArrayDeserializer.java) just to handle similar cases).

So, my assumption that `sizes` won't be empty for images is wrong, we need to fix this in both platforms then, we should:
1. Either gracefully handle the switch to empty array (this is what I'll do in Android).
2. Remove the property from the Networking model, this is an easier solution if we don't use the `sizes` property in the app.

hichamboushaba commented:
An update on why this is happening on your site @selanthiraiyan, one of the images in your site is pretty small (4x2 pixels...), when an image is small, WordPress [skips](https://github.com/WordPress/WordPress/blob/261373653c190d4c77a677f9df75f40a89c4648e/wp-admin/includes/image.php#L79C4-L79C68) generating any other alternative sizes (as they are generally smaller than the original one), so this leads to the empty object.

selanthiraiyan reviewed (approved):
👋  @hafizrahman I am approving this PR as Hicham's explanation makes sense to me. 

Could you please clone this https://github.com/woocommerce/woocommerce-android/issues/12989 for iOS and merge this PR? Thanks!

hafizrahman commented:
> Could you please clone this https://github.com/woocommerce/woocommerce-android/issues/12989 for iOS and merge this PR? Thanks!

Sure: #14516

selanthiraiyan commented:
@hafizrahman The release note for this PR is mentioned under version `21.2`. The PR is targeting `21.3`, though.

hafizrahman commented:
@selanthiraiyan good catch, thanks, fixing in https://github.com/woocommerce/woocommerce-ios/pull/14521/
